crates/graphics/src/collision/instance.rs

```diff
@@ -0,0 +1,204 @@
+        let mut vertices = Vec::with_capacity(width * height * 12);
+
+        for (y, x) in (0..height).cartesian_product(0..width) {
+            let rect =
+                egui::Rect::from_min_size(egui::pos2(x as f32, y as f32), egui::vec2(32., 32.));
+            let center = glam::vec3(rect.center().x, rect.center().y, 0.);
+            let top_left = glam::vec3(rect.left_top().x, rect.left_top().y, 0.);
+            let top_right = glam::vec3(rect.right_top().x, rect.right_top().y, 0.);
+            let bottom_left = glam::vec3(rect.left_bottom().x, rect.left_bottom().y, 0.);
+            let bottom_right = glam::vec3(rect.right_bottom().x, rect.right_bottom().y, 0.);
+
+            vertices.push(Vertex {
+                position: center,
+                direction: 1,
+            });
+            vertices.push(Vertex {
+                position: bottom_left,
+                direction: 1,
+            });
+            vertices.push(Vertex {
+                position: bottom_right,
+                direction: 1,
+            });
+
+            vertices.push(Vertex {
+                position: center,
+                direction: 2,
+            });
+            vertices.push(Vertex {
+                position: top_left,
+                direction: 2,
+            });
+            vertices.push(Vertex {
+                position: bottom_left,
+                direction: 2,
+            });
+
+            vertices.push(Vertex {
+                position: center,
+                direction: 4,
+            });
+            vertices.push(Vertex {
+                position: bottom_right,
+                direction: 4,
+            });
+            vertices.push(Vertex {
+                position: top_right,
+                direction: 4,
+            });
+
+            vertices.push(Vertex {
+                position: center,
+                direction: 8,
+            });
+            vertices.push(Vertex {
+                position: top_right,
+                direction: 8,
+            });
+            vertices.push(Vertex {
+                position: top_left,
+                direction: 8,
+            });
+        }
+
+        vertices
+    }
+
+    pub fn draw<'rpass>(&'rpass self, render_pass: &mut wgpu::RenderPass<'rpass>) {
+        render_pass.set_vertex_buffer(0, self.vertex_buffer.slice(..));
+
+        // Calculate the start and end index of the buffer, as well as the amount of instances.
+        let start_index = 0;
+        let end_index = self.map_width * self.map_height;
+        let count = (end_index - start_index) as u32;
+
+        // Convert the indexes into actual offsets.
+        let start = (start_index * std::mem::size_of::<Instance>()) as wgpu::BufferAddress;
+        let end = (end_index * std::mem::size_of::<Instance>()) as wgpu::BufferAddress;
+
+        render_pass.set_vertex_buffer(1, self.instance_buffer.slice(start..end));
+
+        render_pass.draw(0..12, 0..count);
```

Comment:
This is only drawing the same 12 vertices, why does `calculate_vertices` create more than 12?

Reply:
I guess I misunderstood somewhere how instancing works.

Reply:
That's fair! Graphics programming is pretty confusing